AsyncDisplayKit/ASMapNode.mm

```diff
@@ -0,0 +1,236 @@
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "ASMapNode.h"
+#import <AsyncDisplayKit/ASInsetLayoutSpec.h>
+#import <AsyncDisplayKit/ASCenterLayoutSpec.h>
+#import <AsyncDisplayKit/ASThread.h>
+
+@interface ASMapNode()
+{
+  ASDN::RecursiveMutex _propertyLock;
+  CGSize _nodeSize;
+  MKMapSnapshotter *_snapshotter;
+  MKMapSnapshotOptions *_options;
+  CGSize _maxSize;
+  NSArray *_annotations;
+}
+@end
+
+@implementation ASMapNode
+
+@synthesize liveMap = _liveMap;
+@synthesize mapSize = _mapSize;
+@synthesize automaticallyReloadsMapImageOnOrientationChange = _automaticallyReloadsMapImageOnOrientationChange;
+@synthesize mapDelegate = _mapDelegate;
+
+- (instancetype)initWithCoordinate:(CLLocationCoordinate2D)coordinate
+{
+  if (!(self = [super init])) {
+    return nil;
+  }
+  self.backgroundColor = ASDisplayNodeDefaultPlaceholderColor();
+  _automaticallyReloadsMapImageOnOrientationChange = YES;
+  _liveMap = NO;
+
+  _options = [[MKMapSnapshotOptions alloc] init];
+  _options.region = MKCoordinateRegionMakeWithDistance(coordinate, 1000, 1000);;
+  _mapImage = [[ASImageNode alloc]init];
+  _mapImage.clipsToBounds = YES;
+  [self addSubnode:_mapImage];
+  _maxSize = self.bounds.size;
+  return self;
+}
+
+- (void)addAnnotations:(NSArray *)annotations
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (annotations.count == 0) {
+    return;
+  }
+  _annotations = [annotations copy];
+  if (annotations.count != _annotations.count && _mapImage.image) {
+    // Redraw
+    [self setNeedsDisplay];
+  }
+}
+
+- (void)setUpSnapshotter
+{
+  if (!_snapshotter) {
+    _options.size = _nodeSize;
+    _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+  }
+}
+
+- (BOOL)isLiveMap
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _liveMap;
+}
+
+- (void)setLiveMap:(BOOL)liveMap
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (liveMap == _liveMap) return;
+  _liveMap = liveMap;
+  liveMap ? [self addLiveMap] : [self removeLiveMap];
+}
+
+- (CGSize)mapSize
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _mapSize;
+}
+
+- (void)setMapSize:(CGSize)mapSize
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (CGSizeEqualToSize(mapSize,_mapSize)) {
+    return;
+  }
+  _mapSize = mapSize;
+  _nodeSize = _mapSize;
+  _automaticallyReloadsMapImageOnOrientationChange = NO;
+  [self setNeedsLayout];
+}
+
+- (BOOL)automaticallyReloadsMapImageOnOrientationChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _automaticallyReloadsMapImageOnOrientationChange;
+}
+
+- (void)setAutomaticallyReloadsMapImageOnOrientationChange:(BOOL)automaticallyReloadsMapImageOnOrientationChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (_automaticallyReloadsMapImageOnOrientationChange == automaticallyReloadsMapImageOnOrientationChange) {
+    return;
+  }
+  _automaticallyReloadsMapImageOnOrientationChange = automaticallyReloadsMapImageOnOrientationChange;
+  
```

Comment:
Can remove the last return in this method before the brace.  Since this method has no side effects, you could also remove the if statement / early return and just unconditionally set the ivar after obtaining the lock.